### PR TITLE
Add command line switch to load JSON

### LINKED ISSUE #27: Have command line parameters take priority over JSON
This is definitely open for discussion. I'm personally a fan of removing JSON configuration files and going back to command line parameters. At least give command line priority over JSON (So you can use both but if there's a conflict command line is used).

### PR DESCRIPTION
**Pull request by wchill**
### Short Description:

Adds a command line switch to load `config.json` or some other JSON file using `--json <configuration file>`. If loaded, any arguments in the JSON file take priority over command line switches.

Implementation is open for discussion, though I think this is the cleanest way since the argument parser already provides defaults for the various switches.
### Changes:
- As stated above.

@OpenPoGo/maintainers

Closes #27.

### CODE REVIEW

**Comment by jtdroste** [comment]
👍 

[Approved with PullApprove](https://pullapprove.com/OpenPoGo/OpenPoGoBot/pull-request/37/?utm_source=github-pr&utm_medium=comment-badge&utm_campaign=OpenPoGo/OpenPoGoBot)

**Comment by meyer9** [comment]
:+1:

[Approved with PullApprove](https://pullapprove.com/OpenPoGo/OpenPoGoBot/pull-request/37/?utm_source=github-pr&utm_medium=comment-badge&utm_campaign=OpenPoGo/OpenPoGoBot)

**Comment by jtdroste** [comment]
👍 

[Approved with PullApprove](https://pullapprove.com/OpenPoGo/OpenPoGoBot/pull-request/37/?utm_source=github-pr&utm_medium=comment-badge&utm_campaign=OpenPoGo/OpenPoGoBot)

**Comment by meyer9** [comment]
:-1:

readability comments

[Rejected with PullApprove](https://pullapprove.com/OpenPoGo/OpenPoGoBot/pull-request/37/?utm_source=github-pr&utm_medium=comment-badge&utm_campaign=OpenPoGo/OpenPoGoBot)

File: pokecli.py [hunk 1]
```diff
@@ -46,106 +46,124 @@
...
+        # No JSON file specified, stick with command line args
+        if config.auth_service not in ['ptc', 'google']:
+            logging.error("Invalid Auth service specified! ('ptc' or 'google')")
+            return None
+
+        if not (config.location or config.location_cache):
+            parser.error("Needs either --use-location-cache or --location.")
+            return None
+
+        if not config.username:
```
Comment (meyer9) [inline pokecli.py]: should be if config.username is None

File: pokecli.py [hunk 2]
```diff
@@ -46,106 +46,124 @@
...
+            config.username = raw_input("Username: ")
+        if not config.password:
```
Comment (meyer9) [inline pokecli.py]: same here

File: pokecli.py [hunk 3]
```diff
@@ -46,106 +46,124 @@
...
+        except Exception as e:
+            logging.error("Error loading {}.".format(config.json))
+            return None
+    else:
+        # No JSON file specified, stick with command line args
+        if config.auth_service not in ['ptc', 'google']:
+            logging.error("Invalid Auth service specified! ('ptc' or 'google')")
+            return None
+
+        if not (config.location or config.location_cache):
```
Comment (meyer9) [inline pokecli.py]: should be if (config.location or config.location_cache) is not None

**Comment by sebastienvercammen** [comment]
Maybe consider switching priorities around? `config.json` for configuration, flags for overrides.

**Comment by MaxLeiter** [comment]
Agree with @sebastienvercammen: commandline should take priority over JSON (why would you use commandline params at all then?)

**Comment by meyer9** [comment]
I agree. Let's do that.

**Comment by wchill** [comment]
I think the best way to implement this is to define a dict with our chosen default values and have all of argparser's arguments default to None. If an argument is None, we will use the appropriate value from JSON if specified; if not in the JSON or no file is loaded, we use defaults.

Will fix later today.

**Comment by meyer9** [comment]
:+1:

[Approved with PullApprove](https://pullapprove.com/OpenPoGo/OpenPoGoBot/pull-request/37/?utm_source=github-pr&utm_medium=comment-badge&utm_campaign=OpenPoGo/OpenPoGoBot)

**Comment by jmeyer2k** [comment]
:+1:

**Comment by MaxLeiter** [comment]
:-1: until priorities are switched and comments are addressed

[Rejected with PullApprove](https://pullapprove.com/OpenPoGo/OpenPoGoBot/pull-request/37/?utm_source=github-pr&utm_medium=comment-badge&utm_campaign=OpenPoGo/OpenPoGoBot)

**Comment by fonix232** [comment]
Also I'd change the option name - make it "--config" (and possibly "-c"?) instead of "--json". The latter is just counter-intuitive.

**Comment by wchill** [comment]
-c is potentially more confusing due to the other bot which uses that argument to set CP thresholds.

File: pokecli.py [hunk 4]
```diff
@@ -45,97 +45,119 @@
...
-    parser.add_argument("-u", "--username", help="Username")
-    parser.add_argument("-p", "--password", help="Password")
-    parser.add_argument("-l", "--location", help="Location")
+                        dest="auth_service")
+    parser.add_argument("-u", "--username", help="Username", dest="username")
+    parser.add_argument("-p", "--password", help="Password", dest="password")
+    parser.add_argument("-l", "--location", help="Location", dest="location")
     parser.add_argument("-lc",
-                        "--location_cache",
+                        "--location-cache",
```
Comment (MaxLeiter) [inline pokecli.py]: why the change from _ to -?

File: pokecli.py [hunk 5]
```diff
@@ -45,97 +45,119 @@
...
         "--walk",
         help="Walk instead of teleport with given speed (meters per second, e.g. 2.5)",
         type=float,
-        default=2.5)
-    parser.add_argument("-c",
-                        "--cp",
-                        help="Set CP less than to transfer(DEFAULT 100)",
+        dest="walk")
+    parser.add_argument("-cp",
+                        "--combat-power",
```
Comment (MaxLeiter) [inline pokecli.py]: should be --combat_power

File: pokecli.py [hunk 6]
```diff
@@ -45,97 +45,119 @@
...
+                        "--combat-points",
```
Comment (MaxLeiter) [inline pokecli.py]: should be --combat_points

File: pokecli.py [hunk 7]
```diff
@@ -45,97 +45,119 @@
...
         help="Set the steps around your initial location(DEFAULT 5 mean 25 cells around your location)",
         type=int,
-        default=50)
+        dest="max_steps")
     parser.add_argument(
         "-it",
-        "--initial_transfer",
-        help="Transfer all pokemon with same ID on bot start, except pokemon with highest CP. It works with -c",
-        action='store_true')
+        "--initial-transfer",
```
Comment (MaxLeiter) [inline pokecli.py]: should be --initial_transfer

**Comment by meyer9** [comment]
:+1:

[Approved with PullApprove](https://pullapprove.com/OpenPoGo/OpenPoGoBot/pull-request/37/?utm_source=github-pr&utm_medium=comment-badge&utm_campaign=OpenPoGo/OpenPoGoBot)

**Comment by MaxLeiter** [comment]
:+1: great to see this merged :)

[Approved with PullApprove](https://pullapprove.com/OpenPoGo/OpenPoGoBot/pull-request/37/?utm_source=github-pr&utm_medium=comment-badge&utm_campaign=OpenPoGo/OpenPoGoBot)
